### ISSUE TITLE
DDP hangs at first AllGather during parameter broadcast with tied weights on Blackwell GPUs (sm_120)

### ISSUE THREAD

**Issue by lakshminarasimmanv** [pytorch/pytorch#165727]
Multi-GPU training with DistributedDataParallel (DDP) hangs indefinitely during the initial parameter broadcast phase when training Seq2Seq models with tied weights on Blackwell GPUs (compute capability 12.0).

The hang occurs at the first NCCL AllGather collective operation (8 bytes) before any training iterations begin.

Symptoms:
  - Both GPUs show 100% utilization, 10% memory usage
  - Process hangs after printing "Starting training!"
  - NCCL successfully initializes ring topology but hangs at first AllGather operation
  - Single-GPU training works perfectly
  - Issue is reproducible across ALL distributed training frameworks (torchrun, Accelerate, DeepSpeed, Ray Train)

  Models affected: Any Seq2Seq model with tied encoder-decoder embeddings (BART, T5, mT5, ByT5, etc.)

  Minimal self-contained reproduction script:

  #!/usr/bin/env python3
  # All necessary imports at the beginning
  import torch
  import torch.distributed as dist
  from transformers import AutoModelForSeq2SeqLM, AutoTokenizer, Seq2SeqTrainer, Seq2SeqTrainingArguments
  from datasets import Dataset

  def main():
      # Initialize DDP
      dist.init_process_group(backend="nccl")
      local_rank = dist.get_rank()
      torch.cuda.set_device(local_rank)

      # Load model with tied weights (encoder and decoder share embeddings)
      model = AutoModelForSeq2SeqLM.from_pretrained("facebook/bart-base")
      tokenizer = AutoTokenizer.from_pretrained("facebook/bart-base")

      # Create minimal dataset
      data = {
          "input_ids": [[1, 2, 3, 4, 5]] * 100,
          "labels": [[1, 2, 3, 4, 5]] * 100,
      }
      train_dataset = Dataset.from_dict(data)

      # Configure training
      training_args = Seq2SeqTrainingArguments(
          output_dir="./test_output",
          max_steps=10,
          per_device_train_batch_size=2,
          ddp_find_unused_parameters=False,
          local_rank=local_rank,
      )

      # Create trainer
      trainer = Seq2SeqTrainer(
          model=model,
          args=training_args,
          train_dataset=train_dataset,
          tokenizer=tokenizer,
      )

      print("Starting training!")  # Hangs at this point indefinitely
      trainer.train()

  if __name__ == "__main__":
      main()

  Run with:
  NCCL_DEBUG=INFO NCCL_DEBUG_SUBSYS=ALL torchrun --nproc_per_node=2 reproduce_blackwell_bug.py

  Observed results:
  Process hangs indefinitely after "Starting training!" with both GPUs at 100% utilization and 10% memory usage. NCCL debug output shows:

  atlas-ai-workstation:972413:972413 [0] NCCL INFO comm 0x7ff50c003c60 rank 0 nranks 2 cudaDev 0 busId 200000 - Init COMPLETE
  atlas-ai-workstation:972414:972414 [1] NCCL INFO comm 0x7fcf0c003c60 rank 1 nranks 2 cudaDev 1 busId 200000 - Init COMPLETE

  [... Ring topology setup succeeds ...]

  atlas-ai-workstation:972413:972413 [0] NCCL INFO AllGather: 8 Bytes -> Algo RING proto LL channel{Lo..Hi}={0..0}

  Log ends here. Process never proceeds beyond this AllGather operation.

  Expected results:
  DDP should successfully broadcast initial model parameters and proceed with training, as it does on older GPU architectures (Ampere, Ada Lovelace, etc.).

  Additional context

  Analysis:
  1. Single-GPU training works perfectly - The issue only affects multi-GPU distributed training
  2. NCCL initialization succeeds - Ring topology is correctly established between GPUs
  3. Hang point identified - First AllGather (8 bytes) during DDP's initial parameter broadcast, NOT during gradient synchronization
  4. Tied weights suspected - Issue may be related to how DDP handles tied/shared parameters on Blackwell architecture
  5. Novel bug - No documented cases found in PyTorch issues #145949 (Blackwell tracking) or #157549 (Blackwell compatibility)

  Frameworks tested (ALL exhibit same hang):
  - torchrun with native DDP
  - HuggingFace Accelerate DDP
  - HuggingFace Accelerate FSDP
  - DeepSpeed ZeRO-2
  - Ray Train

  Workarounds attempted:
  - ✗ FSDP instead of DDP
  - ✗ DeepSpeed ZeRO parameter partitioning
  - ✗ Disabling ddp_find_unused_parameters
  - ✗ Different batch sizes and learning rates
  - ✗ Multiple NCCL environment variable configurations

  Impact:
  This bug blocks all multi-GPU training for Seq2Seq models with tied weights on Blackwell GPUs, affecting popular models like BART, T5, mT5, ByT5, and similar architectures.

  cc: @pytorch/distributed

### Versions

PyTorch version: 2.9.0+cu128
Is debug build: False
CUDA used to build PyTorch: 12.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.3 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.39

Python version: 3.13.7 (main, Sep  2 2025, 14:21:46) [Clang 20.1.4 ] (64-bit runtime)
Python platform: Linux-6.14.0-33-generic-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: 12.8.61
CUDA_MODULE_LOADING set to:
GPU models and configuration:
GPU 0: NVIDIA RTX PRO 6000 Blackwell Max-Q Workstation Edition
GPU 1: NVIDIA RTX PRO 6000 Blackwell Max-Q Workstation Edition

Nvidia driver version: 580.65.06
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           52 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  48
On-line CPU(s) list:                     0-47
Vendor ID:                               AuthenticAMD
Model name:                              AMD Ryzen Threadripper PRO 7965WX 24-Cores
CPU family:                              25
Model:                                   24
Thread(s) per core:                      2
Core(s) per socket:                      24
Socket(s):                               1
Stepping:                                1
Frequency boost:                         enabled
CPU(s) scaling MHz:                      59%
CPU max MHz:                             5364.0000
CPU min MHz:                             545.0000
BogoMIPS:                                8387.48
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good amd_lbr_v2 nopl xtopology nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba perfmon_v2 ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local user_shstk avx512_bf16 clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin cppc arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic vgif x2avic v_spec_ctrl vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid overflow_recov succor smca fsrm flush_l1d debug_swap
Virtualization:                          AMD-V
L1d cache:                               768 KiB (24 instances)
L1i cache:                               768 KiB (24 instances)
L2 cache:                                24 MiB (24 instances)
L3 cache:                                128 MiB (4 instances)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-47
Vulnerability Gather data sampling:      Not affected
Vulnerability Ghostwrite:                Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Mitigation; Safe RET
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; STIBP always-on; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsx async abort:           Not affected

Versions of relevant libraries:
[pip3] numpy==2.3.4
[pip3] nvidia-cublas-cu12==12.8.4.1
[pip3] nvidia-cuda-cupti-cu12==12.8.90
[pip3] nvidia-cuda-nvrtc-cu12==12.8.93
[pip3] nvidia-cuda-runtime-cu12==12.8.90
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.3.3.83
[pip3] nvidia-curand-cu12==10.3.9.90
[pip3] nvidia-cusolver-cu12==11.7.3.90
[pip3] nvidia-cusparse-cu12==12.5.8.93
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.27.5
[pip3] nvidia-nvjitlink-cu12==12.8.93
[pip3] nvidia-nvtx-cu12==12.8.90
[pip3] torch==2.9.0
[pip3] triton==3.5.0
[conda] Could not collect
